### PR TITLE
Browser target for SFrame.explore

### LINKED ISSUE #2240: SFrame.explore should support browser target

### PR DESCRIPTION
**Pull request by znation**
* Adds support for the `browser` target for SFrame explore (table view).
* Adds proper data handling on the server, using POST requests and
  `boost::property_tree` for JSON decoding.
* Supports scrolling, jump to row, and accordion expansion.
* Refactors the table spec & data functions out of unity_sframe into
  visualization/server/table.
* Refactors empty_tz construction out of unity_sframe and into static
  local initialization within the escaping functions.
* Refactors image hover to load over HTTP from localhost, for both GUI
  and browser target explore. This appears to be much faster than the
  server queue/push-based approach, since now the client handles its own
  request queue and the server can respond on multiple threads.

Fixes #2240

### CODE REVIEW

File: src/core/storage/sframe_interface/unity_sframe.cpp [hunk 1]
```diff
@@ -1832,323 +1831,29 @@ void unity_sframe::explore(const std::string& path_to_client, const std::string&
...
-    };
+    // Get a reader just once.
+    auto reader = underlying_sframe->get_reader();
 
-    // pass the first 1k rows
-    getRows(0, 100);
+    ew << visualization::table_data(self, reader.get(), 0, 100);
 
-    const size_t resized_height = 200;
+    //const size_t resized_height = 200;
```
Comment (fareeha) [inline src/core/storage/sframe_interface/unity_sframe.cpp]: Should this line be removed?

File: src/core/storage/sframe_interface/unity_sframe.cpp [hunk 2]
```diff
@@ -1832,323 +1831,29 @@ void unity_sframe::explore(const std::string& path_to_client, const std::string&
...
-        DASSERT_LT(index, self->size());
-        DASSERT_GE(index, 0);
+    std::stringstream table_spec;
+    table_spec << "{\"table_spec\":" << visualization::table_spec(self, title) << "}" << std::endl;
+    ew << table_spec.str();
 
-        auto accordion_sa = self->select_column(column_name);
-        auto gl_sa = gl_sarray(accordion_sa);
+    // This materializes if not already
+    auto underlying_sframe = self->get_underlying_sframe();
```
Comment (TobyRoseman) [inline src/core/storage/sframe_interface/unity_sframe.cpp]: `auto` -> `sframe`

File: src/core/storage/sframe_interface/unity_sframe.cpp (hunk 1 above)
Comment (znation) [reply src/core/storage/sframe_interface/unity_sframe.cpp]: Yes, good catch.

File: src/core/storage/sframe_interface/unity_sframe.cpp (hunk 2 above)
Comment (znation) [reply src/core/storage/sframe_interface/unity_sframe.cpp]: Thanks, good catch. I think I was lazily copy/pasting.

**Comment by znation** [comment]
> Is there any way we can add unit tests for any of this?

Good question. I'll look into it. We don't have any infrastructure yet (tracked by https://github.com/apple/turicreate/issues/2153) but it might not be too hard to use something built into Python or already in the Python ecosystem, now that our web UI is exposed in a regular browser (not just custom GUI), and the protocol is regular HTTP (not just custom IPC).

**Comment by znation** [comment]
After some initial investigation, it looks like it'll be non-trivial to set up a browser test environment and some package that allows us to automate the browser. For now I'd like to merge this PR without unit test coverage, and I have issues tracking:

* #2153 - web front-end test infrastructure
* #2247 - SFrame.explore should have UI tests
* #2248 - SFrame.show should have UI tests

I'll also see if there is anything on the C++ side we can unit test independently.

**Pull request merged**